mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClusteringModel.scala

```diff
@@ -0,0 +1,126 @@
+
+import breeze.linalg.{DenseVector => BDV, Vector => BV, norm => breezeNorm}
+import org.apache.spark.Logging
+import org.apache.spark.api.java.JavaRDD
+import org.apache.spark.mllib.linalg.Vector
+import org.apache.spark.rdd.RDD
+
+/**
+ * this class is used for the model of the hierarchical clustering
+ *
+ * @param clusterTree a cluster as a tree node
+ * @param isTrained if the model has been trained, the flag is true
+ */
+class HierarchicalClusteringModel private (
+  val clusterTree: ClusterTree,
+  private[mllib] var isTrained: Boolean) extends Serializable with Logging with Cloneable {
+
+  def this(clusterTree: ClusterTree) = this(clusterTree, false)
+
+  override def clone(): HierarchicalClusteringModel = {
+    new HierarchicalClusteringModel(this.clusterTree.clone(), true)
+  }
+
+  /**
+   * Cuts a cluster tree by given threshold of dendrogram height
+   *
+   * @param height a threshold to cut a cluster tree
+   * @return a hierarchical clustering model
+   */
+  def cut(height: Double): HierarchicalClusteringModel = {
+    val cloned = this.clone()
+    cloned.clusterTree.cut(height)
+    cloned
+  }
+
+  /**
+   * Predicts the closest cluster of each point
+   */
+  def predict(vector: Vector): Int = {
+    // TODO Supports distance metrics other Euclidean distance metric
+    val metric = (bv1: BV[Double], bv2: BV[Double]) => breezeNorm(bv1 - bv2, 2.0)
+    this.clusterTree.assignClusterIndex(metric)(vector)
+  }
+
+  /**
+   * Predicts the closest cluster of each point
+   */
+  def predict(data: RDD[Vector]): RDD[(Int, Vector)] = {
+    val startTime = System.currentTimeMillis() // to measure the execution time
+    val sc = data.sparkContext
+
+    // TODO Supports distance metrics other Euclidean distance metric
+    val metric = (bv1: BV[Double], bv2: BV[Double]) => breezeNorm(bv1 - bv2, 2.0)
+    val treeRoot = this.clusterTree
+    sc.broadcast(metric)
+    sc.broadcast(treeRoot)
+    val predicted = data.map(point => (treeRoot.assignClusterIndex(metric)(point), point))
+
+    val predictTime = System.currentTimeMillis() - startTime
+    logInfo(s"Predicting Time: ${predictTime.toDouble / 1000} [sec]")
+
+    predicted
+  }
+
+  /** Maps given points to their cluster indices. */
+  def predict(points: JavaRDD[Vector]): JavaRDD[java.lang.Integer] =
+    predict(points.rdd).map(_._1).toJavaRDD().asInstanceOf[JavaRDD[java.lang.Integer]]
+
+  /**
+   * Computes the sum of total variance of all cluster
+   */
+  def getSumOfVariance(): Double = this.getClusters().map(_.getVariance().get).sum
+
+  def getClusters(): Array[ClusterTree] = clusterTree.getClusters().toArray
+
+  def getCenters(): Array[Vector] = getClusters().map(_.center)
+
+  /**
+   * Converts a clustering merging list
+   * Returned data format is fit for scipy's dendrogram function
+   * SEE ALSO: scipy.cluster.hierarchy.dendrogram
+   *
+   * @return List[(node1, node2, distance, tree size)]
+   */
+  def toMergeList(): List[(Int, Int, Double, Int)] = {
```

Comment:
Consider renaming -> `toLinkageMatrix`? I think that's a more general term for this data structure. Would require renaming here and elsewhere (e.g. in the Python code).